### PR TITLE
Safety API changes and Cross Platform Tests

### PR DESCRIPTION
**Pull request by Si1ver**
This PR changes some API calls and reimplements tests using different framework.

It also fixes `CWE-416 Use After Free` vulnerability.

### API type safety changes
`trudpChannelData.td` changed from `void*` to correct pointer type. `trudpData` is forward declared for that.
```
typedef struct trudpChannelData {
...
-    void *td; ///< Pointer to trudpData
+    struct trudpData *td; ///< Pointer to trudpData
```

Now one can use `tcd->td->fd` instead of unsafe cast macro `TD(tcd)->fd`.

This also allowed to use implicit pointer type in `trudpChannelNew` for `trudpData`
```
-TRUDP_API trudpChannelData *trudpChannelNew(void *td, char *remote_address, int remote_port_i, int channel);
+TRUDP_API trudpChannelData *trudpChannelNew(struct trudpData *td, char *remote_address, int remote_port_i, int channel);
```

`trudpChannelProcessReceivedPacket` might return pointer to freed memory. Return value was never used, so I decided to remove it.
Additional parameter `data_length` that was used to indicate length of returned data is also removed.
```
-TRUDP_API void *trudpChannelProcessReceivedPacket(trudpChannelData *tcd, uint8_t *data, size_t packet_length, size_t *data_length);
+TRUDP_API int trudpChannelProcessReceivedPacket(trudpChannelData *tcd, uint8_t *data, size_t packet_length);
```

New return value is int.
 - 1 when Trudp packet is handled;
 - 0 if received data is not a Trudp packet;
 - -1 if invalid Trudp packet received.

The return value of `trudpSendEventGotData` was never used. I desided to remove it.
Additional parameter `data_length` that was used to indicate length of returned data is also removed.
This function is internal to Trudp so it now marked accordingly.
```
-void *trudpSendEventGotData(trudpChannelData* tcd, trudpPacket *packet, size_t *data_length);
+TRUDP_INTERNAL void trudpSendEventGotData(trudpChannelData *tcd, trudpPacket *packet);
```

### Test changes
Tests rewritten with different test framework called `Cheat` http://users.jyu.fi/~sapekiis/cheat/
Pros:
 - Single file test framework (no dependencies).
 - Cross-platform support (tests now run on windows).
 - Less code needed to write new test.
 
Cons:
 - All tests must be implemented in single file.
 - Preprocessor is heavily involved and it's harder to debug tests.

### CODE REVIEW

**Comment by Si1ver** [comment]
I added cheat.h to CodeFactor ignore list because it's third party library that's hard to change.

File: src/trudp_channel.c [hunk 1]
```diff
@@ -589,21 +589,16 @@ size_t trudpChannelSendData(trudpChannelData *tcd, void *data,
  * @param tcd Pointer to trudpChannelData
  * @param packet Pointer to received packet
  * @param packet_length Packet length
- * @param data_length Pointer to variable to return packets data length
  *
- * @return Pointer to received data, NULL available, length of data saved to
- *         *data_length, if packet is not TR-UDP packet it's sent to client as
- * is
+ * @return 1 if packet is processed, 0 id packet is not a trudp packet, -1 on
```
Comment (chernikovdmitry) [inline src/trudp_channel.c]:
```suggestion
 * @return 1 if packet is processed, 0 if packet is not a trudp packet, -1 on
```

File: src/trudp.c [hunk 2]
```diff
@@ -262,10 +252,8 @@ void trudpProcessReceived(trudpData* td, uint8_t* data, size_t data_length) {
 
     // Process received packet
     if(recvlen > 0) {
-        size_t packet_data_length;
-
         trudpChannelData *tcd = trudpGetChannelCreate(td, (__CONST_SOCKADDR_ARG) &remaddr, 0);
-        if(tcd == (void *)-1 || trudpChannelProcessReceivedPacket(tcd, data, recvlen, &packet_data_length) == (void *)-1) {
+        if(tcd == (void *)-1 || trudpChannelProcessReceivedPacket(tcd, data, recvlen) == -1) {
```
Comment (chernikovdmitry) [inline src/trudp.c:258]:
```suggestion
        // FIXME non trudp data it's return value == 0, not -1. Investigate why it works and fix appropriately
        if(tcd == (void *)-1 || trudpChannelProcessReceivedPacket(tcd, data, recvlen) == -1) {
```

File: tests/cheat/cheat.h [hunk 3]
```diff
@@ -0,0 +1,3313 @@
...
+This is free software, and you are welcome to redistribute it
+under certain conditions; see the LICENSE file for details.
+*/
+
+/*
+Identifiers starting with CHEAT_ and cheat_ are reserved for internal use and
+identifiers starting with cheat_test_, cheat_wrapped_ or cheat_unwrapped_ for
+external use.
+*/
+
```
Comment (chernikovdmitry) [inline tests/cheat/cheat.h:14]:
```suggestion

/*
    Original version 1.0.4
*/

```

**Review by chernikovdmitry — CHANGES_REQUESTED** [review]
Few one-line changes

**Review by kirill-scherba — APPROVED** [review]

**Pull request merged**